File: SmartDeviceLink_Example/Classes/Preferences.m
//
//  Preferences.m
//  SmartDeviceLink-iOS

#import "Preferences.h"


NSString *const IPAddressPreferencesKey = @"SDLExampleAppIPAddress";
NSString *const PortPreferencesKey = @"SDLExampleAppPort";

NSString *const DefaultIPAddressValue = @"192.168.1.1";
UInt16 const DefaultPortValue = 12345;



@interface Preferences ()

@end



@implementation Preferences


#pragma mark - Singleton / Initializers

+ (instancetype)sharedPreferences {
    static Preferences *sharedPreferences = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedPreferences = [[Preferences alloc] init];
    });
    
    return sharedPreferences;
}

- (instancetype)init {
    self = [super init];
    if (!self) {
        return nil;
    }
    
    if (self.ipAddress == nil || self.port == 0) {
        [self resetPreferences];
    }
    
    return self;
}


#pragma mark - Public API

- (void)resetPreferences {
    self.ipAddress = DefaultIPAddressValue;
    self.port = DefaultPortValue;
}


#pragma mark - Setters / Getters

- (NSString *)ipAddress {
    return [self stringForKey:IPAddressPreferencesKey];
}

- (void)setIpAddress:(NSString *)ipAddress {
    [self setString:ipAddress forKey:IPAddressPreferencesKey];
}

- (UInt16)port {
    return (UInt16)[self integerForKey:PortPreferencesKey];
}

- (void)setPort:(UInt16)port {
    [self setInteger:port forKey:PortPreferencesKey];
}


#pragma mark - Private User Defaults Helpers

- (void)setString:(NSString *)aString forKey:(NSString *)aKey {
    NSParameterAssert(aKey != nil);
    
    dispatch_async(self.class.preferencesQueue, ^{
        [[NSUserDefaults standardUserDefaults] setObject:aString forKey:aKey];
    });
}

- (NSString *)stringForKey:(NSString *)aKey {
    NSParameterAssert(aKey != nil);
    
    __block NSString *retVal = nil;
    dispatch_sync(self.class.preferencesQueue, ^{
        retVal = [[NSUserDefaults standardUserDefaults] stringForKey:aKey];
    });
    
    return retVal;
}

- (void)setInteger:(NSInteger)aInt forKey:(NSString *)aKey {
    NSParameterAssert(aKey != nil);
    
    dispatch_async(self.class.preferencesQueue, ^{
        [[NSUserDefaults standardUserDefaults] setInteger:aInt forKey:aKey];
    });
}

- (NSInteger)integerForKey:(NSString *)aKey {
    NSParameterAssert(aKey != nil);
    
    __block NSInteger retVal = 0;
    dispatch_sync(self.class.preferencesQueue, ^{
        retVal = [[NSUserDefaults standardUserDefaults] integerForKey:aKey];
    });
    
    return retVal;
}


#pragma mark - Class Queue

+ (dispatch_queue_t)preferencesQueue {
    static dispatch_queue_t preferencesQueue;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        preferencesQueue = dispatch_queue_create("com.sdl-example.preferences", DISPATCH_QUEUE_SERIAL);
    });
    
    return preferencesQueue;
}

@end

File: SmartDeviceLink-ExampleSwift/Preferences.swift
//
//  Preferences.swift
//  SmartDeviceLink-iOS
//
//  Created by Brett McIsaac on 5/15/17.
//  Copyright © 2017 smartdevicelink. All rights reserved.
//

import Foundation

var ipAddress: String = ""
var port: UInt16 = 0
let IPAddressPreferencesKey: String = "SDLExampleAppIPAddress"
let PortPreferencesKey: String = "SDLExampleAppPort"
let DefaultIPAddressValue: String = "192.168.1.1"
let DefaultPortValue: UInt16 = 12345

class Preferences {
    // MARK: - Singleton / Initializers
    
    override init() {
        super.init()
        if ipAddress == nil || port == 0 {
            reset()
        }
    }
    
    class func shared() -> Preferences {
        var sharedPreferences: Preferences? = nil
        var onceToken: Int
        if (onceToken == 0) {
            /* TODO: move below code to a static variable initializer (dispatch_once is deprecated) */
            sharedPreferences = Preferences()
        }
        onceToken = 1
        return sharedPreferences!
    }
    
    // MARK: - Class Queue
    class func preferencesQueue() -> DispatchQueue {
        var preferencesQueue: DispatchQueue
        var onceToken: Int
        if (onceToken == 0) {
            /* TODO: move below code to a static variable initializer (dispatch_once is deprecated) */
            preferencesQueue = DispatchQueue(label: "com.sdl-exampleSwift.preferences")
        }
        onceToken = 1
        return preferencesQueue
    }
    
    // MARK: - Public API
    func resetPreferences() {
        ipAddress = DefaultIPAddressValue
        port = DefaultPortValue
    }
    
    // MARK: - Setters / Getters
    
    // MARK: - Private User Defaults Helpers
    func set(_ aString: String, forKey aKey: String) {
        assert(aKey != nil, "Invalid parameter not satisfying: aKey != nil")
        Preferences.preferencesQueue().async(execute: {() -> Void in
            UserDefaults.standard.set(aString, forKey: aKey)
        })
    }
    
    func string(forKey aKey: String) -> String {
        assert(aKey != nil, "Invalid parameter not satisfying: aKey != nil")
        var retVal: String? = nil
        Preferences.preferencesQueue().sync(execute: {() -> Void in
            retVal = UserDefaults.standard.string(forKey: aKey)
        })
        return retVal!
    }
    
    func set(_ aInt: Int, forKey aKey: String) {
        assert(aKey != nil, "Invalid parameter not satisfying: aKey != nil")
        Preferences.preferencesQueue().async(execute: {() -> Void in
            UserDefaults.standard.set(aInt, forKey: aKey)
        })
    }
    
    func integer(forKey aKey: String) -> Int {
        assert(aKey != nil, "Invalid parameter not satisfying: aKey != nil")
        var retVal: Int = 0
        Preferences.preferencesQueue().sync(execute: {() -> Void in
            retVal = UserDefaults.standard.integer(forKey: aKey)
        })
        return retVal
    }
}
